File: bindings/go-open/dhruv/engine.go
package dhruv

import (
	"fmt"
	"runtime"

	"ctara-dhruv-core/bindings/go-open/internal/cabi"
)

type Engine struct {
	h cabi.EngineHandle
}

type LSK struct {
	h cabi.LskHandle
}

type EOP struct {
	h cabi.EopHandle
}

type Config struct {
	h cabi.ConfigHandle
}

func APIVersion() uint32 {
	return cabi.APIVersion()
}

func VerifyABI() error {
	if got := APIVersion(); got != ExpectedAPIVersion {
		return fmt.Errorf("verify_abi failed: expected ABI v%d, got v%d", ExpectedAPIVersion, got)
	}
	return nil
}

func LoadConfig(path string) (*Config, error) {
	h, st := cabi.LoadConfig(path)
	if err := statusErr("config_load", st); err != nil {
		return nil, err
	}
	cfg := &Config{h: h}
	runtime.SetFinalizer(cfg, (*Config).Close)
	return cfg, nil
}

func ClearActiveConfig() error {
	return statusErr("config_clear_active", cabi.ConfigClearActive())
}

func (c *Config) Close() error {
	if c == nil {
		return nil
	}
	runtime.SetFinalizer(c, nil)
	return statusErr("config_free", c.h.Free())
}

func NewEngine(cfg EngineConfig) (*Engine, error) {
	h, st, prior := cabi.CreateEngine(cfg)
	if err := wrapStatus("engine_new", st, prior); err != nil {
		return nil, err
	}
	e := &Engine{h: h}
	runtime.SetFinalizer(e, (*Engine).Close)
	return e, nil
}

func (e *Engine) Close() error {
	if e == nil {
		return nil
	}
	runtime.SetFinalizer(e, nil)
	return statusErr("engine_free", e.h.Free())
}

func (e *Engine) Query(q QueryRequest) (QueryResult, error) {
	out, st := cabi.QueryEngineRequest(e.h, q)
	return out, statusErr("engine_query", st)
}

func QueryOnce(cfg EngineConfig, q Query) (StateVector, error) {
	out, st, prior := cabi.QueryOnce(cfg, q)
	return out, wrapStatus("query_once", st, prior)
}

func LoadLSK(path string) (*LSK, error) {
	h, st := cabi.LoadLSK(path)
	if err := statusErr("lsk_load", st); err != nil {
		return nil, err
	}
	l := &LSK{h: h}
	runtime.SetFinalizer(l, (*LSK).Close)
	return l, nil
}

func (l *LSK) Close() error {
	if l == nil {
		return nil
	}
	runtime.SetFinalizer(l, nil)
	return statusErr("lsk_free", l.h.Free())
}

func LoadEOP(path string) (*EOP, error) {
	h, st := cabi.LoadEOP(path)
	if err := statusErr("eop_load", st); err != nil {
		return nil, err
	}
	e := &EOP{h: h}
	runtime.SetFinalizer(e, (*EOP).Close)
	return e, nil
}

func (e *EOP) Close() error {
	if e == nil {
		return nil
	}
	runtime.SetFinalizer(e, nil)
	return statusErr("eop_free", e.h.Free())
}

func CartesianToSpherical(position [3]float64) (SphericalCoords, error) {
	out, st := cabi.CartesianToSpherical(position)
	return out, statusErr("cartesian_to_spherical", st)
}

File: bindings/node-open/src/engine.js
'use strict';

const { addon } = require('./native');
const { EXPECTED_API_VERSION } = require('./status');
const { checkStatus } = require('./errors');

class Config {
  constructor(handle) {
    this._handle = handle;
    this._closed = false;
  }

  static load(path = null, defaultsMode = 0) {
    const r = addon.configLoad(path, defaultsMode);
    checkStatus('config_load', r.status);
    return new Config(r.handle);
  }

  close() {
    if (this._closed) return;
    checkStatus('config_free', addon.configFree(this._handle));
    this._closed = true;
    this._handle = null;
  }
}

class LSK {
  constructor(handle) {
    this._handle = handle;
    this._closed = false;
  }

  static load(path) {
    const r = addon.lskLoad(path);
    checkStatus('lsk_load', r.status);
    return new LSK(r.handle);
  }

  close() {
    if (this._closed) return;
    checkStatus('lsk_free', addon.lskFree(this._handle));
    this._closed = true;
    this._handle = null;
  }
}

class EOP {
  constructor(handle) {
    this._handle = handle;
    this._closed = false;
  }

  static load(path) {
    const r = addon.eopLoad(path);
    checkStatus('eop_load', r.status);
    return new EOP(r.handle);
  }

  close() {
    if (this._closed) return;
    checkStatus('eop_free', addon.eopFree(this._handle));
    this._closed = true;
    this._handle = null;
  }
}

class Engine {
  constructor(handle) {
    this._handle = handle;
    this._closed = false;
  }

  static create(config) {
    const r = addon.engineNew(config);
    checkStatus('engine_new', r.status);
    return new Engine(r.handle);
  }

  close() {
    if (this._closed) return;
    checkStatus('engine_free', addon.engineFree(this._handle));
    this._closed = true;
    this._handle = null;
  }

  query(query) {
    const r = addon.engineQuery(this._handle, query);
    checkStatus('engine_query', r.status);
    return r.result;
  }

  replaceSpks(spkPaths) {
    const r = addon.engineReplaceSpks(this._handle, { spkPaths });
    checkStatus('engine_replace_spks', r.status);
    return r.report;
  }

  listSpks() {
    const r = addon.engineListSpks(this._handle);
    checkStatus('engine_list_spks', r.status);
    return r.spks;
  }
}

function apiVersion() {
  return addon.apiVersion();
}

function verifyAbi() {
  const got = apiVersion();
  if (got !== EXPECTED_API_VERSION) {
    throw new Error(`ABI mismatch: expected ${EXPECTED_API_VERSION}, got ${got}`);
  }
}

function clearActiveConfig() {
  checkStatus('config_clear_active', addon.configClearActive());
}

function queryOnce(config, query) {
  const r = addon.queryOnce(config, query);
  checkStatus('query_once', r.status);
  return r.state;
}

function cartesianToSpherical(positionKm) {
  const r = addon.cartesianToSpherical(positionKm);
  checkStatus('cartesian_to_spherical', r.status);
  return r.coords;
}

module.exports = {
  Config,
  Engine,
  EOP,
  LSK,
  apiVersion,
  verifyAbi,
  clearActiveConfig,
  queryOnce,
  cartesianToSpherical,
};
